### PR TITLE
miscellaneous fixes

### PR DESCRIPTION
**Pull request by tycho**
## Description
I've been maintaining a somewhat large patch set for my game and am hoping to get a bunch of these changes upstreamed so that I can re-enable the "dynapi" feature (since I'm basically binary incompatible with SDL 2.x right now). There are two different branches involved in this, and this is the simpler of the two.

These are miscellaneous fixes that probably don't warrant individual pull requests for each (or maybe they do, let me know). The commit messages are pretty clear about what is going on in each.

## Existing Issue(s)
None

### CODE REVIEW

File: src/video/cocoa/SDL_cocoawindow.m [hunk 1]
```diff
@@ -1738,13 +1738,19 @@ - (BOOL)acceptsFirstMouse:(NSEvent *)theEvent
...
+
 #if MAC_OS_X_VERSION_MAX_ALLOWED >= 101200 /* Added in the 10.12.0 SDK. */
     /* By default, don't allow users to make our window tabbed in 10.12 or later */
     if ([nswindow respondsToSelector:@selector(setTabbingMode:)]) {
         [nswindow setTabbingMode:NSWindowTabbingModeDisallowed];
     }
 #endif
 
+    if ([nswindow respondsToSelector:@selector(setReleasedWhenClosed:)]) {
+        [nswindow setReleasedWhenClosed:NO];
```
Comment (slime73) [inline src/video/cocoa/SDL_cocoawindow.m:1751]: This is already done in SDL's code: https://github.com/libsdl-org/SDL/blob/49d6f3e6c3aaa1596dd2cc770baa3dd85d4ad2a4/src/video/cocoa/SDL_cocoawindow.m#L1685

File: src/video/cocoa/SDL_cocoawindow.m [hunk 2]
```diff
@@ -1738,13 +1738,19 @@ - (BOOL)acceptsFirstMouse:(NSEvent *)theEvent
         return SDL_SetError("%s", [[e reason] UTF8String]);
     }
 
+    [nswindow setColorSpace:[NSColorSpace sRGBColorSpace]];
```
Comment (slime73) [inline src/video/cocoa/SDL_cocoawindow.m:1741]: This caused a big performance drop on older macOS versions (we had it in there for like an hour but backed it out: https://github.com/libsdl-org/SDL/commit/fd85f5748db4852c3ee5f8fc8253c691c7ca89fc). I haven't observed a similar performance drop in macOS 12 with it, but at the very least I think this should be heavily tested and also gated behind a system version check.

File: src/video/cocoa/SDL_cocoawindow.m (hunk 1 above)
Comment (tycho) [reply src/video/cocoa/SDL_cocoawindow.m:1751]: Aha, I've rebased so many times I missed that, so thanks for catching that. I'll drop that patch.

File: src/video/cocoa/SDL_cocoawindow.m (hunk 2 above)
Comment (tycho) [reply src/video/cocoa/SDL_cocoawindow.m:1741]: Hmm, okay. I'll try to figure out how to make an OS version check to make it only happen on macOS 12+, perhaps.

Comment (slime73) [reply src/video/cocoa/SDL_cocoawindow.m:1741]: I've only tested it recently on a discrete AMD GPU so I'm not confident yet that it's cost-free on all GPU setups.

Comment (tycho) [reply src/video/cocoa/SDL_cocoawindow.m:1741]: My tests have been on AMD GPUs and Apple M1/M2. Not sure there's anything else still supported by Apple to care about.

Comment (slime73) [reply src/video/cocoa/SDL_cocoawindow.m:1741]: Many Macs have integrated Intel GPUs.

File: src/video/windows/SDL_windowswindow.c [hunk 3]
```diff
@@ -117,6 +124,20 @@ GetWindowStyle(SDL_Window * window)
...
 
+static void
+LoadDWMFunctions()
+{
+    static BOOL triedOnce = FALSE;
+    void *pDwmApiDll = NULL;
+    if (triedOnce)
+        return;
+    triedOnce = TRUE;
+    pDwmApiDll = SDL_LoadObject("dwmapi.dll");
```
Comment (slime73) [inline src/video/windows/SDL_windowswindow.c:135]: This will remain loaded after the video subsystem is quit/deinitialized, right? There's a convention in SDL_windowsvideo.c to load and unload referenced winapi DLLs at appropriate times.

File: src/video/windows/SDL_windowswindow.c [hunk 4]
```diff
@@ -518,6 +539,14 @@ WIN_CreateWindow(_THIS, SDL_Window * window)
         return WIN_SetError("Couldn't create window");
     }
 
+    if (SDL_GetHintBoolean(SDL_HINT_WINDOWS_DWM_DARK_MODE, SDL_FALSE)) {
+        BOOL value = TRUE;
+        if (!pDwmSetWindowAttribute)
+            LoadDWMFunctions();
```
Comment (slouken) [inline src/video/windows/SDL_windowswindow.c:545]: It's probably not worth creating a separate function in SDL_windowswindow.c to load just one function. Go ahead and do the loadobject/loadfunction inline here, as we do a bunch of other places.

File: src/video/windows/SDL_windowsevents.c [hunk 5]
```diff
@@ -1267,7 +1267,14 @@ WIN_WindowProc(HWND hwnd, UINT msg, WPARAM wParam, LPARAM lParam)
 
         /* We'll do our own drawing, prevent flicker */
     case WM_ERASEBKGND:
+        if (!data->videodata->cleared)
```
Comment (slouken) [inline src/video/windows/SDL_windowsevents.c:1270]: What is the interaction with an application that starts with a white loading screen?

File: src/video/cocoa/SDL_cocoawindow.m (hunk 2 above)
Comment (slouken) [reply src/video/cocoa/SDL_cocoawindow.m:1741]: Is this maybe a bug in the Metal rendering, using sRGB when we should be using a different colorspace? I don't think we expect an sRGB colorspace in any of the SDL rendering code.

File: src/video/windows/SDL_windowsevents.c (hunk 5 above)
Comment (tycho) [reply src/video/windows/SDL_windowsevents.c:1270]: Maybe I should make it a hint then. Every game I've ever done starts with black, but I can see the rationale.

But the other annoying thing here is that Windows has been wildly inconsistent about what it paints the window with. Even if you set a specific brush for it to use on the first draw in the CreateWindow(Ex?) function, Windows seems to ignore that value these days, and will just fill it white unless you do something special in `WM_ERASEBKGND`.

File: src/video/windows/SDL_windowswindow.c (hunk 3 above)
Comment (tycho) [reply src/video/windows/SDL_windowswindow.c:135]: Yeah, I should change that.

I might be able to just do `SDL_LoadObjact`, `SDL_LoadFunction`, call `DWMSetWindowAttribute`, `SDL_UnloadObject` and the attribute would likely stay applied properly anyway, so no need to leak the handle.

File: src/video/cocoa/SDL_cocoawindow.m (hunk 2 above)
Comment (slime73) [reply src/video/cocoa/SDL_cocoawindow.m:1741]: This is mostly an OpenGL-specific change/fix actually, as far as I know: on Macs with P3 displays, the backbuffer colors of the opengl content view in the window are treated as 0-1 P3 instead of 0-1 (linear or encoded) sRGB, because the compositor just copies the backbuffer data to the display.

So without this change, OpenGL apps have oversaturated colors on P3 displays because the system is effectively reading P3 colorspace data from the window contents when the app didn't intend it.

Comment (tycho) [reply src/video/cocoa/SDL_cocoawindow.m:1741]: I'm not sure what colorspace OpenGL core is using on macOS internally but it's doing something close to Display P3, even though Apple asserts that their OpenGL implementation is colorspace agnostic. It ends up causing the colors to look a lot more saturated than they probably should (based on MoltenVK/Metal behavior, and comparison with another non-Mac system).

File: src/video/windows/SDL_windowsevents.c (hunk 5 above)
Comment (slime73) [reply src/video/windows/SDL_windowsevents.c:1270]: For what it's worth, SDL intentionally unconditionally clears the window contents to black on macOS when the window is created: https://github.com/libsdl-org/SDL/blob/main/src/video/cocoa/SDL_cocoawindow.m#L1557

**Review by jabundis13 — APPROVED** [review]

File: include/SDL_hints.h [hunk 6]
```diff
@@ -2048,6 +2048,20 @@ extern "C" {
...
+ *  By default, Windows does not opt any applications into adhering to the dark
+ *  mode color scheme for title bars/borders. This hint tells Windows to use
+ *  the appropriate dark/light mode colors for created SDL windows.
+ *
+ *  This variable can be set to the following values:
+ *    "0"       - Windows will always use the light mode color scheme.
+ *    "1"       - Windows will follow the user light/dark mode preference and use
+ *                dark title bar and window borders when appropriate.
+ */
+#define SDL_HINT_WINDOWS_DWM_DARK_MODE "SDL_WINDOWS_DWM_DARK_MODE"
```
Comment (cgutman) [inline include/SDL_hints.h:2063]: I wonder if we should make a cross-platform hint for this instead of making it Windows-only.

File: include/SDL_hints.h [hunk 7]
```diff
@@ -2048,6 +2048,20 @@ extern "C" {
...
+ * \brief Opts created windows into the Windows DWM dark mode appearance.
+ *
+ *  By default, Windows does not opt any applications into adhering to the dark
+ *  mode color scheme for title bars/borders. This hint tells Windows to use
+ *  the appropriate dark/light mode colors for created SDL windows.
+ *
+ *  This variable can be set to the following values:
+ *    "0"       - Windows will always use the light mode color scheme.
+ *    "1"       - Windows will follow the user light/dark mode preference and use
+ *                dark title bar and window borders when appropriate.
```
Comment (cgutman) [inline include/SDL_hints.h:2061]: Based on the name of the hint, I would expect to always get dark mode if I set this. If this is only opting in to follow the system light/dark preference, I think we should use a name that describes behavior specifically.

File: include/SDL_hints.h (hunk 6 above)
Comment (tycho) [reply include/SDL_hints.h:2063]: Maybe, though I don't know what the implementation would look like for any other platforms.

File: include/SDL_hints.h (hunk 7 above)
Comment (tycho) [reply include/SDL_hints.h:2061]: Yeah, maybe something like `SDL_HINT_WINDOWS_HONOR_DARK_MODE` or something, not sure. But that's getting a bit wordy. Ideas?

Comment (slouken) [reply include/SDL_hints.h:2061]: Silly question, is there a case where we _don't_ want to honor the OS preference? I'm thinking we just add dark mode support and follow the OS preference for all platforms, and if we need to add a hint, it's to disable that behavior.

File: src/video/windows/SDL_windowsevents.c (hunk 5 above)
Comment (slouken) [reply src/video/windows/SDL_windowsevents.c:1270]: You're right. I checked a game that I've seen with a white loading screen, and the initial window is black before switching to a white loading screen.

File: include/SDL_hints.h (hunk 7 above)
Comment (slouken) [reply include/SDL_hints.h:2075]: Can you please split this out into a separate PR? I'd like to merge it later once we add dark mode support (possibly in SDL3)

**Comment by slouken** [comment]
Okay, I cherry-picked all the fixes except the dark mode hint. Can you create a separate PR for that?

Thanks!
